toolchain/parser/parser_impl.cpp

```diff
@@ -1180,6 +1180,87 @@ auto ParseTree::Parser::ParseWhileStatement() -> llvm::Optional<Node> {
                  /*has_error=*/!cond || !body);
 }
 
+auto ParseTree::Parser::ParseForStatement() -> llvm::Optional<Node> {
+  // TODO Add proper diagnostics.
+  CARBON_RETURN_IF_STACK_LIMITED(llvm::None);
+  auto start = GetSubtreeStartPosition();
+  auto token = Consume(TokenKind::For());
+
+  auto header = [this]() -> llvm::Optional<Node> {
+    auto open_paren = ConsumeIf(TokenKind::OpenParen());
+    auto header_start = GetSubtreeStartPosition();
+
+    if (!open_paren) {
+      CARBON_DIAGNOSTIC(ExpectedParenAfter, Error, "Expected `(` after `{0}`.",
+                        TokenKind);
+      emitter_.Emit(*position_, ExpectedParenAfter, TokenKind::For());
+    } else {
+      // Question: should we:
+      // 1. Try to continue parsing the loop header?
+      // 2. Skip all tokens until we reach either:
+      //    * A `)` without an opening one?
+      //      That might be easy because:
+      //      - The lexer will likely add a recovery token and mark the `)` as
+      //        an error token. So we skip until we come across this error
+      //        token.
+      //    * A `{` which marks the start of the block:
+      //      We might want to implement a heuristic here that the `{` has to be
+      //      the last token on the line.
+      //    * The end of line.
+      //
+      // The first option boils down to trying to report as many errors as
+      // possible in the header. The second one tries to skip over as much as
+      // possible until we reach a new construct, i.e. the for body block.
+    }
+
+    bool var_decl_parsed = false;
+
+    if (NextTokenIs(TokenKind::Var())) {
+      auto var_token = Consume(TokenKind::Var());
+      auto var_start = GetSubtreeStartPosition();
+      auto pattern = ParsePattern(PatternKind::Variable);
+      AddNode(ParseNodeKind::VariableDeclaration(), var_token, var_start,
+              !pattern);
+      var_decl_parsed = true;
+    } else {
+      CARBON_DIAGNOSTIC(ExpectedVariableDeclaration, Error,
+                        "Expected `var` declaration.");
+      emitter_.Emit(*position_, ExpectedVariableDeclaration);
+
+      // Same questin for missing `(` applies here.
```

Comment:
Scanning forward looking for the `in` or `)` seems reasonable if pattern parsing fails.